### ISSUE TITLE
Rework attributes to be more flexible

### ISSUE THREAD

**Issue by jl-wynen** [scipp/scipp#2675]
This came up in the past (#2213):

> Do we need xarray-style attrs, e.g., to represent attributes from HDF5 files?
>
>  - Should make decision now, even if not implement, since renaming current attrs property will become harder and harder.
>  - Are our attrs sufficient, such that they could be used for HDF5-style attrs (provided that we add attr support to Variable and Dataset)?
>      -  Should not have any mechanism that restricts names of attributes... but currently coordinate names would clash and cause problems.
>  - Are there any concerns apart from finding a better name for current attrs?
       - Performance should not be a big issue, can use nullptr if no attributes, i.e., only minor size overhead. May want avoid/reduce use of PyObject to avoid GIL.
>  - Potential solution: Remove attrs, use meta to access all meta data, allow insertion into meta (inserts unaligned coord, i.e., not into coords). => Opens path to adding HDF5-style attrs later.

Is the fact that we have been silent on this issue basically a decision for not doing it? It has already become more difficult to do it than when we first talked about it.

Opening this issue now because I came across a case where such attributes would be useful. Let's keep a list of use cases:
-  Store a bootstrap history in a data array. But the dimension of the history does not match any of the data array's dims.
- The majority of attributes in data loaded from nexus files are either scalar strings or scalars wrapping data arrays.

**Comment by SimonHeybrock** [comment]
> Store a bootstrap history in a data array. But the dimension of the history does not match any of the data array's dims.

Any reason why this cannot be wrapper in a scalar variable?

**Comment by jl-wynen** [comment]
> Any reason why this cannot be wrapper in a scalar variable?

Technically, no. Everything could be stored that way.

**Comment by jl-wynen** [comment]
> My question is where you see the advantage in not doing that?

It simplifies the syntax, you no longer need to write `.value` everywhere.

But thinking about mapping to HDF5 attributes, storing a large object like this in such an attribute would be a bad idea. Not sure if this is relevant though, since we are not trying to mimic HDF5.

**Comment by SimonHeybrock** [comment]
There is one spot of bother in scipp that surfaces occasionally: `sc.Dataset` does not support attrs and thus `ds['x', 5]` changes the `x` coord into dataset *item* attributes. The reason for this is as follows:

- Originally we had dataset and dataset-item (data array) attributes. This caused complication and conceptual issues from shadowing.
- Dataset-items must support attributes to store metadata (such as sample information, ...).
- Therefore, dataset attr support was removed.

Suppose we added a separate mechanism to handle "non-axis" meta data.For clarity, let us give everything a name (this is not a proposal, just for clarity of discussion):

- `coords` -> `aligned_coords`
- `attrs` -> `unaligned_coords` (some attrs)
- `attrs` -> `metadata` (most attrs)

Then we could have:
```
da.aligned_coords
da.unaligned_coords  # aligned turns into this when slicing
da.metadata  # mostly non-coord stuff, such as sample info, ...

ds.aligned_coords
ds.unaligned_coords  # aligned turns into this when slicing
ds.metadata  # maybe, maybe not
ds['a'].metadata
ds['b'].metadata
ds['a'].aligned_coords  # selection of ds.aligned_coords
ds['a'].unaligned_coords  # selection of ds.unaligned_coords
```
Note how *all items in `ds` have the same `unaligned_coords`. That is, `Dataset` *does* require alignment of these coords (unlike operations). The difference to the current state is that the things that are always unaligned are no longer living in `attrs` but in `metadata`.

The result of this would be a saner slicing behavior of `Dataset`, and `transform_coords` would be nicer (currently it has to deal with attrs of all items separately).
